aiida_mlip/data/model.py

```diff
@@ -0,0 +1,125 @@
+"""Define Model Data type in AiiDA"""
+
+import hashlib
+from pathlib import Path
+from urllib import request
+from urllib.parse import urlparse
+
+from aiida.orm import SinglefileData
+
+
+def calculate_hash(file):
+    """Function to calculate file hashes"""
+    # Calculate hash
+    sha256 = hashlib.sha256()
+    with open(file, "rb") as f:
+        for data in iter(lambda: f.read(65536), b""):
+            sha256.update(data)
+    file_hash = sha256.hexdigest()
+    return file_hash
+
+
+def check_existing_file(arch_path, file, file_hash, architecture, cache_path):
+    """Function to check if a file already exists and return a file path and architecture"""
+    for existing_file in arch_path.rglob("*"):
+        if existing_file.is_file() and existing_file != file:
+            existing_file_hash = calculate_hash(existing_file)
+            if existing_file_hash == file_hash:
+                ex_file_path = Path(existing_file).parent
+                if arch_path == ex_file_path:
+                    file.unlink()
+                    return existing_file, architecture
+
+                if arch_path == cache_path and arch_path != ex_file_path:
+                    architecture = ex_file_path.name
+                    file.unlink()
+                    return existing_file, architecture
+    return file, architecture
+
+
+class ModelData(SinglefileData):
+    """Class to handle models"""
+
+    def __init__(self, file, filename, architecture: str = None, **kwargs):
+        """Add a file to the node, parse it and set the attributes found.
+
+        :param file: absolute path to the file or a filelike object
+        :param filename: specify filename to use (defaults to name of provided file).
```

Comment:
It may be more obvious when I actually use this, but I'm not sure it's as clear as it could be what the difference between `file` and `filepath` is.

Reply:
basically file is the path to the file, while filename is the name you want to give in aiida to save that model.
so if the path to the file is `folder/file1` and you give filename = None, filename will be `file1`. But if you give another name it willuse that to define the file and save it in the database
I called them this way cause this is how they were called in the SinglefileData class, but I also think it's not too intuitive. So I guess I coulkd change it, call them like filepath and modelname or something

Reply:
Ah ok, I think matching names with `SinglefileData` is definitely valid, even if the names aren't perfect - consistency is always useful.

Perhaps if you could just elaborate slightly in the docstring (even just adding something similar to `... object whose contents to copy` as is in `SinglefileData` for `file` might be enough)?